core/src/main/java/org/apache/carbondata/core/dictionary/generator/TableDictionaryGenerator.java

```diff
@@ -115,7 +115,10 @@ public Integer size(DictionaryMessage key) {
   }
 
   public void updateGenerator(CarbonDimension dimension) {
-    columnMap.put(dimension.getColumnId(),
-            new IncrementalColumnDictionaryGenerator(dimension, 1));
+    // reuse dictionary generator
+    if (null == columnMap.get(dimension.getColumnId())) {
+      columnMap.put(dimension.getColumnId(),
```

Comment:
lock is required to be taken on columnMap. with double checking, otherwise still parallel intialization is possible.